internal/batches/executor/coordinator.go

```diff
@@ -0,0 +1,269 @@
+		AutoAuthorDetails: opts.AutoAuthorDetails,
+		Parallelism:       opts.Parallelism,
+		Timeout:           opts.Timeout,
+		TempDir:           opts.TempDir,
+	})
+
+	return &Coordinator{
+		opts: opts,
+
+		cache:      cache,
+		exec:       exec,
+		logManager: logManager,
+	}
+}
+
+// CheckCache checks whether the internal ExecutionCache contains
+// ChangesetSpecs for the given Tasks. If cached ChangesetSpecs exist, those
+// are returned, otherwise the Task, to be executed later.
+func (c *Coordinator) CheckCache(ctx context.Context, tasks []*Task) (uncached []*Task, specs []*batches.ChangesetSpec, err error) {
+	for _, t := range tasks {
+		cachedSpecs, found, err := c.checkCacheForTask(ctx, t)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if !found {
+			uncached = append(uncached, t)
+			continue
+		}
+
+		specs = append(specs, cachedSpecs...)
+	}
+
+	return uncached, specs, nil
+}
+
+func (c *Coordinator) checkCacheForTask(ctx context.Context, task *Task) (specs []*batches.ChangesetSpec, found bool, err error) {
+	// Check if the task is cached.
+	cacheKey := task.cacheKey()
+	if c.opts.ClearCache {
+		if err := c.cache.Clear(ctx, cacheKey); err != nil {
+			return specs, false, errors.Wrapf(err, "clearing cache for %q", task.Repository.Name)
+		}
+
+		return specs, false, nil
+	}
+
+	var result executionResult
+	result, found, err = c.cache.Get(ctx, cacheKey)
+	if err != nil {
+		return specs, false, errors.Wrapf(err, "checking cache for %q", task.Repository.Name)
+	}
+
+	if !found {
+		return specs, false, nil
+	}
+
+	// If the cached result resulted in an empty diff, we don't need to
+	// add it to the list of specs that are displayed to the user and
+	// send to the server. Instead, we can just report that the task is
+	// complete and move on.
+	if result.Diff == "" {
+		return specs, true, nil
+	}
+
+	specs, err = createChangesetSpecs(task, result, c.opts.AutoAuthorDetails)
+	if err != nil {
+		return specs, false, err
+	}
+
+	return specs, true, nil
+}
+
+func (c *Coordinator) cacheAndBuildSpec(ctx context.Context, taskResult taskResult, status taskStatusHandler) ([]*batches.ChangesetSpec, error) {
+	// Add to the cache, even if no diff was produced.
+	cacheKey := taskResult.task.cacheKey()
+	if err := c.cache.Set(ctx, cacheKey, taskResult.result); err != nil {
+		return nil, errors.Wrapf(err, "caching result for %q", taskResult.task.Repository.Name)
+	}
+
+	// If the steps didn't result in any diff, we don't need to create a
+	// changeset spec that's displayed to the user and send to the server.
+	if taskResult.result.Diff == "" {
+		return nil, nil
+	}
+
+	// Build the changeset specs.
+	specs, err := createChangesetSpecs(taskResult.task, taskResult.result, c.opts.AutoAuthorDetails)
+	if err != nil {
+		return specs, err
+	}
+
+	// Update the status of Task
+	status.Update(taskResult.task, func(status *TaskStatus) {
+		status.ChangesetSpecs = specs
+		// TODO(mrnugget): Ideally we'd get rid of this dependency on the task
```

Comment:
The reason we still have this dependency is that we log information about the diff in the TUI.